internal/utils/gateway/ownerrefs.go

```
@@ -104,3 +109,85 @@ func ListControlPlanesForGateway(
+// that owns the controlplane by means of the controlplane's owner reference
+func getGatewayFromControlPlane(ctx context.Context,
+	c client.Client,
+	controlplane *operatorv1alpha1.ControlPlane,
+) (*gatewayv1alpha2.Gateway, error) {
+	gateway := &gatewayv1alpha2.Gateway{}
+
+	expectedOwnRef := &metav1.OwnerReference{}
+	expectedOwnRef.Kind = "Gateway"
+	expectedOwnRef.APIVersion = gatewayv1alpha2.SchemeGroupVersion.String()
+
+	for _, ownRef := range controlplane.OwnerReferences {
+		if ownRef.Kind == expectedOwnRef.Kind || ownRef.APIVersion == expectedOwnRef.APIVersion {
+			if err := c.Get(ctx, types.NamespacedName{Namespace: controlplane.Namespace, Name: ownRef.Name}, gateway); err != nil {
+				return nil, err
+			}
+			return gateway, nil
+		}
+	}
+
+	return nil, gatewayerrors.NewOwnerReferenceNotFound(*expectedOwnRef, controlplane.DeepCopy())
+}
+
+// GetDataplaneServiceForControlplane is a helper functions that retrieves
+// the dataplane service owned by dataplane associated to the controlplane
+func GetDataplaneServiceForControlplane(
+	ctx context.Context,
+	c client.Client,
+	controlplane *operatorv1alpha1.ControlPlane,
+) (*corev1.Service, error) {
+
+	dataplane := operatorv1alpha1.DataPlane{}
+	if controlplane.Spec.DataPlane != nil && *controlplane.Spec.DataPlane != "" {
+		dataplaneName := *controlplane.Spec.DataPlane
+		if err := c.Get(ctx, types.NamespacedName{Namespace: controlplane.Namespace, Name: dataplaneName}, &dataplane); err != nil {
+			return nil, err
+		}
+	} else {
+		gateway, err := getGatewayFromControlPlane(ctx, c, controlplane)
+		if err != nil {
+			return nil, err
+		}
+
+		dataplanes, err := ListDataPlanesForGateway(ctx, c, gateway)
+		if err != nil {
+			return nil, err
+		}
+
+		count := len(dataplanes)
+		if count > 1 {
+			return nil, fmt.Errorf("data plane deployments found: %d, expected: 1, requeing", count)
+		}
+		if count == 0 {
+			return nil, fmt.Errorf("no dataplanes found")
+		}
+		dataplane = dataplanes[0]
+	}
```

I don't think we need that else block. In controplane has no dataplane set we should just return an error

---

Since `dataplane` is an optional field, I don't think we should return an error when no dataplane is set in the controlplane. From my point of view, having an optional field that triggers errors when not set, doesn't make much sense.
https://github.com/Kong/gateway-operator/blob/main/api/v1alpha1/controlplane_types.go#L87-L88

---

Hmm, maybe, I get what you're saying. We should clarify that. @shaneutt

---

It's technically valid for a `ControlPlane` to be created without a `DataPlane` to associate it with, in fact we're simultaneously adding "kindness" functionality around this in #51. However it seems worth noting that if we don't have a dataplane associated with the `ControlPlane` in the spec, there can be no associated `Service` so this else block _would_ seem unnecessary: we should return with an indication that there is no `Service`. I think indicating this condition with an `error` is legitimate.

---

My assumption in writing this else block was about having a controlplane with no `spec.dataPlaneName` set, but with a gateway in place. Now that I figured out this situation can never happen, I removed it. Note: I replaced the `onerReferenceNotFoundError`, which doesn't make sense anymore, with `DataPlaneNotSetError`, which is the actual situation we need to cope with.